File: src/api/Models/Borehole.cs
using BDMS.Json;
using NetTopologySuite.Geometries;
using System.ComponentModel.DataAnnotations.Schema;

namespace BDMS.Models;

/// <summary>
/// Represents a borehole entity in the database.
/// </summary>
[Table("borehole")]
public class Borehole : IChangeTracking, IIdentifyable
{
    /// <inheritdoc />
    [IncludeInExport]
    [Column("id")]
    public int Id { get; set; }

    /// <inheritdoc />
    [Column("creator")]
    public int? CreatedById { get; set; }

    /// <inheritdoc />
    public User? CreatedBy { get; set; }

    /// <inheritdoc />
    [IncludeInExport]
    [Column("creation")]
    public DateTime? Created { get; set; }

    /// <inheritdoc />
    [IncludeInExport]
    [Column("update")]
    public DateTime? Updated { get; set; }

    /// <inheritdoc />
    [Column("updater")]
    public int? UpdatedById { get; set; }

    /// <inheritdoc />
    public User? UpdatedBy { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/> locked date.
    /// </summary>
    [Column("locked")]
    public DateTime? Locked { get; set; }

    /// <summary>
    /// Gets or sets the id of the <see cref="User"/> who locked the <see cref="Borehole"/>.
    /// </summary>
    [Column("locked_by")]
    public int? LockedById { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="User"/> who locked the <see cref="Borehole"/>.
    /// </summary>
    public User? LockedBy { get; set; }

    /// <summary>
    /// Gets or sets the foreign key to the <see cref="Workgroup"/> entity.
    /// </summary>
    [Column("workgroup_id")]
    public int? WorkgroupId { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Workgroup"/>.
    /// </summary>
    public Workgroup? Workgroup { get; set; }

    /// <summary>
    /// Gets or sets whether the <see cref="Borehole"/> is public.
    /// </summary>
    [IncludeInExport]
    [Column("public")]
    public bool? IsPublic { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s type id.
    /// </summary>
    [IncludeInExport]
    [Column("borehole_type_id")]
    public int? TypeId { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s type.
    /// </summary>
    public Codelist? Type { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s X-location using LV95 coordinates.
    /// </summary>
    [IncludeInExport]
    [Column("location_x")]
    public double? LocationX { get; set; }

    /// <summary>
    /// Gets or sets the precision for the <see cref="Borehole"/>'s X-location (LV95).
    /// </summary>
    [IncludeInExport]
    [Column("precision_location_x")]
    public int? PrecisionLocationX { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s Y-location using LV95 coordinates.
    /// </summary>
    [IncludeInExport]
    [Column("location_y")]
    public double? LocationY { get; set; }

    /// <summary>
    /// Gets or sets the precision for the <see cref="Borehole"/>'s Y-location (LV95).
    /// </summary>
    [IncludeInExport]
    [Column("precision_location_y")]
    public int? PrecisionLocationY { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s X-location using LV03 coordinates.
    /// </summary>
    [IncludeInExport]
    [Column("location_x_lv03")]
    public double? LocationXLV03 { get; set; }

    /// <summary>
    /// Gets or sets the precision for the <see cref="Borehole"/>'s X-location (LV03).
    /// </summary>
    [IncludeInExport]
    [Column("precision_location_x_lv03")]
    public int? PrecisionLocationXLV03 { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s Y-location using LV03 coordinates.
    /// </summary>
    [IncludeInExport]
    [Column("location_y_lv03")]
    public double? LocationYLV03 { get; set; }

    /// <summary>
    /// Gets or sets the precision for the <see cref="Borehole"/>'s Y-location (LV03).
    /// </summary>
    [IncludeInExport]
    [Column("precision_location_y_lv03")]
    public int? PrecisionLocationYLV03 { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s original reference system id.
    /// </summary>
    [IncludeInExport]
    [Column("srs_id")]
    public int? OriginalReferenceSystemId { get; set; }

    public Codelist? OriginalReferenceSystem { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s Elevation(Z).
    /// </summary>
    [IncludeInExport]
    [Column("elevation_z")]
    public double? ElevationZ { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s HrsId.
    /// </summary>
    [IncludeInExport]
    [Column("hrs_id")]
    public int? HrsId { get; set; }

    public Codelist? Hrs { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s total depth.
    /// </summary>
    [IncludeInExport]
    [Column("total_depth")]
    public double? TotalDepth { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s restriction id.
    /// </summary>
    [IncludeInExport]
    [Column("restriction_id")]
    public int? RestrictionId { get; set; }

    public Codelist? Restriction { get; set; }

    /// <summary>
    /// Gets or sets the date until when a <see cref="Borehole"/> is restricted.
    /// </summary>
    [IncludeInExport]
    [Column("restriction_until")]
    public DateOnly? RestrictionUntil { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s national interest.
    /// </summary>
    [IncludeInExport]
    [Column("national_interest")]
    public bool? NationalInterest { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s original name.
    /// </summary>
    [IncludeInExport]
    [Column("original_name")]
    public string? OriginalName { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s alternate name.
    /// </summary>
    [IncludeInExport]
    [Column("name")]
    public string? Name { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s location precision.
    /// </summary>
    [IncludeInExport]
    [Column("precision_location_id")]
    public int? LocationPrecisionId { get; set; }

    public Codelist? LocationPrecision { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s elevation precision id.
    /// </summary>
    [IncludeInExport]
    [Column("precision_elevation_id")]
    public int? ElevationPrecisionId { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s elevation precision.
    /// </summary>
    public Codelist? ElevationPrecision { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s project name.
    /// </summary>
    [IncludeInExport]
    [Column("project_name")]
    public string? ProjectName { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s country.
    /// </summary>
    [IncludeInExport]
    [Column("country")]
    public string? Country { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s canton.
    /// </summary>
    [IncludeInExport]
    [Column("canton")]
    public string? Canton { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s municipality.
    /// </summary>
    [IncludeInExport]
    [Column("municipality")]
    public string? Municipality { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s purpose id.
    /// </summary>
    [IncludeInExport]
    [Column("purpose_id")]
    public int? PurposeId { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s purpose.
    /// </summary>
    public Codelist? Purpose { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s status id.
    /// </summary>
    [IncludeInExport]
    [Column("status_id")]
    public int? StatusId { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s status.
    /// </summary>
    public Codelist? Status { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s Depth presicion id.
    /// </summary>
    [IncludeInExport]
    [Column("precision_depth_id")]
    public int? DepthPrecisionId { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s Depth presicion.
    /// </summary>
    public Codelist? DepthPrecision { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s top bedrock.
    /// </summary>
    [IncludeInExport]
    [Column("top_bedrock_fresh_md")]
    public double? TopBedrockFreshMd { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s weathered top bedrock.
    /// </summary>
    [IncludeInExport]
    [Column("top_bedrock_weathered_md")]
    public double? TopBedrockWeatheredMd { get; set; }

    /// <summary>
    /// Gets or sets whether the <see cref="Borehole"/> has groundwater.
    /// </summary>
    [IncludeInExport]
    [Column("groundwater")]
    public bool? HasGroundwater { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s geometry.
    /// </summary>
    [IncludeInExport]
    [Column("geometry")]
    public Point? Geometry { get; set; }

    /// <summary>
    /// Gets or sets remarks for  the <see cref="Borehole"/>.
    /// </summary>
    [IncludeInExport]
    [Column("remarks")]
    public string? Remarks { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s lithology top bedrock id.
    /// </summary>
    [IncludeInExport]
    [Column("lithology_top_bedrock_id")]
    public int? LithologyTopBedrockId { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s lithology top bedrock.
    /// </summary>
    public Codelist? LithologyTopBedrock { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s lithostratigraphy top bedrock id.
    /// </summary>
    [IncludeInExport]
    [Column("lithostratigraphy_top_bedrock_id")]
    public int? LithostratigraphyTopBedrockId { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s lithostratigraphy top bedrock.
    /// </summary>
    public Codelist? LithostratigraphyTopBedrock { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s chronostratigraphy top bedrock id.
    /// </summary>
    [IncludeInExport]
    [Column("chronostratigraphy_top_bedrock_id")]
    public int? ChronostratigraphyTopBedrockId { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s chronostratigraphy top bedrock.
    /// </summary>
    public Codelist? ChronostratigraphyTopBedrock { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s reference elevation.
    /// </summary>
    [IncludeInExport]
    [Column("reference_elevation")]
    public double? ReferenceElevation { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s reference elevation precision id.
    /// </summary>
    [IncludeInExport]
    [Column("precision_reference_elevation_id")]
    public int? ReferenceElevationPrecisionId { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s reference elevation precision.
    /// </summary>
    public Codelist? ReferenceElevationPrecision { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s reference elevation type id.
    /// </summary>
    [IncludeInExport]
    [Column("reference_elevation_type_id")]
    public int? ReferenceElevationTypeId { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s true vertical total depth.
    /// </summary>
    [NotMapped]
    public double? TotalDepthTvd { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s true vertical top bedrock fresh depth.
    /// </summary>
    [NotMapped]
    public double? TopBedrockFreshTvd { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s true vertical top bedrock weathered depth.
    /// </summary>
    [NotMapped]
    public double? TopBedrockWeatheredTvd { get; set; }

    /// <summary>
    /// Gets or sets whether <see cref="Borehole"/>'s top bedrock was intersected.
    /// </summary>
    [IncludeInExport]
    [Column("top_bedrock_intersected")]
    public bool? TopBedrockIntersected { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s reference elevation type.
    /// </summary>
    public Codelist? ReferenceElevationType { get; set; }

    /// <summary>
    /// Gets or sets the <see cref="Borehole"/>'s workflow.
    /// </summary>
    public Workflow? Workflow { get; set; }

    /// <summary>
    /// Gets the <see cref="Borehole"/>'s stratigraphies.
    /// </summary>
    [IncludeInExport]
    public ICollection<Stratigraphy>? Stratigraphies { get; set; }

    /// <summary>
    /// Gets the <see cref="Borehole"/>'s completions.
    /// </summary>
    [IncludeInExport]
    public ICollection<Completion>? Completions { get; set; }

    /// <summary>
    /// Gets the <see cref="Borehole"/>'s <see cref="Section"/>s.
    /// </summary>
    [IncludeInExport]
    public ICollection<Section>? Sections { get; set; }

    /// <summary>
    /// Gets the <see cref="Borehole"/>'s <see cref="BoreholeGeometry"/>.
    /// </summary>
    [IncludeInExport]
    public IList<BoreholeGeometryElement>? BoreholeGeometry { get; set; }

    /// <summary>
    /// Gets the <see cref="Borehole"/>'s observations.
    /// </summary>
    [IncludeInExport]
    public ICollection<Observation>? Observations { get; set; }

    /// <summary>
    /// Gets the <see cref="Profile"/>s attached to the <see cref="Borehole"/>.
    /// </summary>
    [IncludeInExport]
    public ICollection<Profile>? Profiles { get; set; }

    /// <summary>
    /// Gets the <see cref="Borehole"/>'s photos.
    /// </summary>
    public ICollection<Photo>? Photos { get; set; }

    /// <summary>
    /// Gets the <see cref="Borehole"/>'s documents.
    /// </summary>
    [IncludeInExport]
    public ICollection<Document>? Documents { get; set; }

    /// <summary>
    /// Gets the <see cref="Borehole"/>'s log runs.
    /// </summary>
    [IncludeInExport]
    public ICollection<LogRun>? LogRuns { get; set; }

    /// <summary>
    /// Gets the <see cref="Codelist"/>s used by the <see cref="Borehole"/> to store multiple id types.
    /// </summary>
    public ICollection<Codelist>? Codelists { get; }

    /// <summary>
    /// Gets the<see cref="BoreholeCodelist"/> join table entities.
    /// </summary>
    [IncludeInExport]
    public IList<BoreholeCodelist>? BoreholeCodelists { get; set; } = new List<BoreholeCodelist>();
}

File: src/client/src/api/borehole.ts
import { GridRowSelectionModel } from "@mui/x-data-grid";
import { keepPreviousData, useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { FeatureCollection, Geometry } from "geojson";
import { Codelist } from "../components/codelist.ts";
import { Photo } from "../pages/detail/attachments/tabs/photo.ts";
import { Observation } from "../pages/detail/form/hydrogeology/Observation.ts";
import { defaultHrsId, referenceSystems } from "../pages/detail/form/location/coordinateSegmentConstants.ts";
import { ReferenceSystemCode } from "../pages/detail/form/location/coordinateSegmentInterfaces.ts";
import { LogRun } from "../pages/detail/form/log/logInterfaces.ts";
import { Workflow } from "../pages/detail/form/workflow/workflow.ts";
import { SessionKeys } from "../pages/overview/SessionKey.ts";
import { Document, NullableDateString, User, Workgroup } from "./apiInterfaces.ts";
import { BoreholeGeometry } from "./boreholeGeometry.ts";
import { Completion } from "./completion.ts";
import { download, downloadData } from "./download.ts";
import { fetchApiV2Legacy, fetchApiV2WithApiError, upload } from "./fetchApiV2.ts";
import { Profile } from "./file/fileInterfaces.ts";
import { Section } from "./section.ts";
import { Stratigraphy } from "./stratigraphy.ts";
import { useCurrentUser } from "./user.ts";

export interface BasicIdentifier {
  id?: number;
  boreholeId: number;
  codelistId: number | null;
  codelist?: Codelist;
  value: string;
  comment?: string | null;
}

// Avoids circular reference for BoreholeV2
export interface Identifier extends BasicIdentifier {
  borehole?: BoreholeV2 | null;
}

export interface BoreholeV2 {
  lithologyTopBedrockId: number;
  lithostratigraphyTopBedrockId: number;
  chronostratigraphyTopBedrockId: number;
  hasGroundwater: boolean | null;
  topBedrockWeatheredMd: number;
  topBedrockFreshMd: number;
  topBedrockIntersected: boolean | null;
  depthPrecisionId: number;
  totalDepth: number;
  purposeId: number;
  typeId: number;
  remarks: string;
  statusId: number;
  workflow: Workflow | null;
  boreholeCodelists: BasicIdentifier[];
  workgroupId: number;
  workgroup: Workgroup;
  originalReferenceSystem: ReferenceSystemCode;
  precisionLocationYLV03: number;
  precisionLocationXLV03: number;
  precisionLocationY: number;
  precisionLocationX: number;
  locationXLV03: number | null;
  locationYLV03: number | null;
  id: number;
  locationX: number | null;
  locationY: number | null;
  municipality: string;
  country: string;
  canton: string;
  name: string;
  originalName: string;
  projectName: number;
  restrictionId: number;
  restrictionUntil: NullableDateString;
  nationalInterest: boolean | null;
  elevationZ: number | string; // Number with thousands separator then parsed to number
  elevationPrecisionId: number;
  referenceElevation: number | string; // Number with thousands separator then parsed to number
  referenceElevationPrecisionId: number;
  referenceElevationTypeId: number;
  locationPrecisionId: number | null;
  hrsId: number;
  updated: NullableDateString;
  updatedById: number;
  updatedBy: User;
  stratigraphies: Stratigraphy[] | null;
  logRuns: LogRun[] | null;
  locked: NullableDateString;
  lockedById: number | null;
  completions: Completion[] | null;
  observations: Observation[] | null;
  sections: Section[] | null;
  boreholeGeometry: BoreholeGeometry[] | null;
  boreholeFiles: Profile[] | null;
  photos: Photo[] | null;
  documents: Document[] | null;
  geometry: BoreholeGeometry | null;
}

const getIdQuery = (ids: number[] | GridRowSelectionModel) => ids.map(id => `ids=${id}`).join("&");

export const exportJsonBoreholes = async (boreholeIds: number[] | GridRowSelectionModel, fileName: string) => {
  const exportJsonResponse = await fetchApiV2Legacy(`boreholeexport/json?${getIdQuery(boreholeIds)}`, "GET");
  const jsonString = JSON.stringify(exportJsonResponse);
  downloadData(jsonString, `${fileName}.json`, "application/json");
};

export const importBoreholesCsv = async (workgroupId: number | null, combinedFormData: FormData) => {
  return await upload(`import/csv?workgroupId=${workgroupId}`, "POST", combinedFormData);
};

export const importBoreholesJson = async (workgroupId: number | null, combinedFormData: FormData) => {
  return await upload(`import/json?workgroupId=${workgroupId}`, "POST", combinedFormData);
};

export const importBoreholesZip = async (workgroupId: number | null, combinedFormData: FormData) => {
  return await upload(`import/zip?workgroupId=${workgroupId}`, "POST", combinedFormData);
};

const createBorehole = async (workgroupId: number): Promise<BoreholeV2> => {
  return await fetchApiV2WithApiError<BoreholeV2>(`borehole`, "POST", {
    workgroupId,
    originalReferenceSystem: referenceSystems.LV95.code,
    hrsId: defaultHrsId,
  });
};

export const copyBorehole = async (boreholeId: GridRowSelectionModel, workgroupId: number | null) => {
  return await fetchApiV2Legacy(`borehole/copy?id=${boreholeId}&workgroupId=${workgroupId}`, "POST");
};

export const exportCSVBorehole = async (boreholeIds: GridRowSelectionModel, fileName: string) => {
  const csvData = await fetchApiV2Legacy(`boreholeexport/csv?${getIdQuery(boreholeIds)}`, "GET");
  downloadData(csvData, `${fileName}.csv`, "text/csv");
};

export const exportJsonWithAttachmentsBorehole = async (boreholeIds: number[] | GridRowSelectionModel) => {
  return await download(`boreholeexport/zip?${getIdQuery(boreholeIds)}`);
};

const fetchBoreholeById = async (id: number): Promise<BoreholeV2> => {
  return await fetchApiV2WithApiError<BoreholeV2>(`borehole/${id}`, "GET");
};

const updateBorehole = async (borehole: BoreholeV2): Promise<BoreholeV2> => {
  return await fetchApiV2WithApiError<BoreholeV2>("borehole", "PUT", borehole);
};
const deleteBorehole = async (id: number) => await fetchApiV2WithApiError(`borehole?id=${id}`, "DELETE");

const canUserEditBorehole = async (id: number) =>
  await fetchApiV2WithApiError<boolean>(`permissions/canedit?boreholeId=${id}`, "GET");

const canUserUpdateBoreholeStatus = async (id: number) =>
  await fetchApiV2WithApiError<boolean>(`permissions/canchangestatus?boreholeId=${id}`, "GET");

export const boreholeQueryKey = "boreholes";

export const useBorehole = (id: number) => {
  return useQuery({
    queryKey: [boreholeQueryKey, id],
    queryFn: async () => {
      return await fetchBoreholeById(id);
    },
    enabled: !!id,
  });
};

export const canEditQueryKey = "canEditBorehole";
export const useBoreholeEditable = (id: number) => {
  const { data: currentUser } = useCurrentUser();
  return useQuery({
    queryKey: [canEditQueryKey, currentUser?.id, id],
    queryFn: async () => {
      return await canUserEditBorehole(id);
    },
    enabled: !!id,
  });
};

export const canUpdateStatusQueryKey = "canUpdateBoreholeStatus";
export const useBoreholeStatusEditable = (id: number) => {
  const { data: currentUser } = useCurrentUser();
  return useQuery({
    queryKey: [canUpdateStatusQueryKey, currentUser?.id, id],
    queryFn: async () => {
      return await canUserUpdateBoreholeStatus(id);
    },
    enabled: !!id,
  });
};

export const useBoreholeMutations = () => {
  const queryClient = useQueryClient();

  const useAddBorehole = useMutation({
    mutationFn: async (workgroupId: number) => {
      return await createBorehole(workgroupId);
    },
  });

  const useUpdateBorehole = useMutation({
    mutationFn: async (borehole: BoreholeV2) => {
      return await updateBorehole(borehole);
    },
    onSuccess: (_, borehole) => {
      queryClient.invalidateQueries({
        queryKey: [boreholeQueryKey],
      });
      // force immediate background refetch to have the borehole's lock status up to date on next render and prevent button flickering
      queryClient.refetchQueries({
        queryKey: [boreholeQueryKey, borehole.id],
        exact: true,
      });
    },
  });

  const useDeleteBorehole = useMutation({
    mutationFn: async (boreholeId: number) => {
      return await deleteBorehole(boreholeId);
    },
    onSuccess: () => {
      queryClient.invalidateQueries({
        queryKey: [boreholeQueryKey],
      });
    },
  });
  return {
    add: useAddBorehole,
    update: useUpdateBorehole,
    delete: useDeleteBorehole,
  };
};

export const useReloadBoreholes = () => {
  const queryClient = useQueryClient();
  return () => {
    queryClient.invalidateQueries({ queryKey: [boreholeQueryKey] });
  };
};

// ---- Filter interfaces ----
export interface BoreholeListItem {
  id: number;
  originalName: string | null;
  name: string | null;
  workgroupId: number | null;
  workgroupName: string | null;
  statusId: number | null;
  typeId: number | null;
  purposeId: number | null;
  restrictionId: number | null;
  restrictionUntil: NullableDateString;
  nationalInterest: boolean | null;
  locationX: number | null;
  locationY: number | null;
  elevationZ: number | null;
  totalDepth: number | null;
  created: NullableDateString;
  updated: NullableDateString;
  isPublic: boolean | null;
  locked: NullableDateString;
}

enum NullableBooleanFilter {
  false,
  true,
  null,
}

enum BooleanFilter {
  false,
  true,
}

type BooleanFilterValue = "true" | "false" | undefined;
type NullableBooleanFilterValue = BooleanFilterValue | "null";

interface BaseFilterRequest {
  polygon?: Geometry | null;
  originalName?: string | null;
  projectName?: string | null;
  name?: string | null;
  statusId?: number[] | null;
  typeId?: number[] | null;
  purposeId?: number[] | null;
  workgroupId?: number[] | null;
  ids?: number[] | null;
  restrictionId?: number[] | null;
  restrictionUntilFrom?: string | null;
  restrictionUntilTo?: string | null;
  totalDepthMin?: number | null;
  totalDepthMax?: number | null;
  topBedrockFreshMdMin?: number | null;
  topBedrockFreshMdMax?: number | null;
  topBedrockWeatheredMdMin?: number | null;
  topBedrockWeatheredMdMax?: number | null;
  pageNumber?: number;
  pageSize?: number;
  orderBy?: string | null;
  direction?: string | null;
  workflowStatus?: string[] | null;
}

export interface FilterRequest extends BaseFilterRequest {
  nationalInterest?: NullableBooleanFilterValue;
  topBedrockIntersected?: NullableBooleanFilterValue;
  hasGroundwater?: NullableBooleanFilterValue;
  hasGeometry?: BooleanFilterValue;
  hasLogs?: BooleanFilterValue;
  hasProfiles?: BooleanFilterValue;
  hasPhotos?: BooleanFilterValue;
  hasDocuments?: BooleanFilterValue;
}

interface FilterRequestSubmission extends BaseFilterRequest {
  nationalInterest?: NullableBooleanFilter;
  topBedrockIntersected?: NullableBooleanFilter;
  hasGroundwater?: NullableBooleanFilter;
  hasGeometry?: BooleanFilter;
  hasLogs?: BooleanFilter;
  hasProfiles?: BooleanFilter;
  hasPhotos?: BooleanFilter;
  hasDocuments?: BooleanFilter;
}

export interface FilterResponse {
  totalCount: number;
  pageNumber: number;
  pageSize: number;
  totalPages: number;
  boreholes: BoreholeListItem[];
  geoJson: FeatureCollection | null;
  filteredBoreholeIds: number[];
  selectableBoreholeIds: number[];
}

export interface NullableBooleanCounts {
  true: number;
  false: number;
  null: number;
}

export interface BooleanCounts {
  true: number;
  false: number;
}

export interface FilterStatsResponse {
  statusId: Record<number, number>;
  typeId: Record<number, number>;
  purposeId: Record<number, number>;
  workgroupId: Record<number, number>;
  restrictionId: Record<number, number>;
  workflowStatusCount: Record<string, number>;
  nationalInterest: NullableBooleanCounts;
  topBedrockIntersected: NullableBooleanCounts;
  hasGroundwater: NullableBooleanCounts;
  hasGeometry: BooleanCounts;
  hasLogs: BooleanCounts;
  hasProfiles: BooleanCounts;
  hasPhotos: BooleanCounts;
  hasDocuments: BooleanCounts;
}

export type BoreholeSuggestionField = "originalName" | "projectName" | "name";

interface BoreholeSuggestion {
  value: string;
  count: number;
}

export const filterBoreholes = async (filterRequest: FilterRequestSubmission): Promise<FilterResponse> => {
  return await fetchApiV2WithApiError<FilterResponse>("borehole/filter", "POST", filterRequest);
};

export const fetchFilterStats = async (filterRequest: FilterRequestSubmission): Promise<FilterStatsResponse> => {
  return await fetchApiV2WithApiError<FilterStatsResponse>("borehole/filter/stats", "POST", filterRequest);
};

const fetchBoreholeSuggestions = (
  field: BoreholeSuggestionField,
  query: string,
  limit = 10,
): Promise<BoreholeSuggestion[]> => {
  const params = new URLSearchParams({
    field,
    query,
    limit: String(limit),
  });
  return fetchApiV2WithApiError<BoreholeSuggestion[]>(`borehole/suggest?${params.toString()}`, "GET");
};

const parseBooleanFilter = (value: "true" | "false" | undefined | null): BooleanFilter | undefined => {
  if (value === "true") return BooleanFilter.true;
  if (value === "false") return BooleanFilter.false;
  return undefined;
};

const parseNullableBooleanFilter = (
  value: "true" | "false" | "null" | undefined | null,
): NullableBooleanFilter | undefined => {
  if (value === "true") return NullableBooleanFilter.true;
  if (value === "false") return NullableBooleanFilter.false;
  if (value === "null") return NullableBooleanFilter.null;
  return undefined;
};

export const toFilterRequestSubmission = (filterRequest: FilterRequest): FilterRequestSubmission => ({
  ...filterRequest,
  hasGroundwater: parseNullableBooleanFilter(filterRequest.hasGroundwater),
  topBedrockIntersected: parseNullableBooleanFilter(filterRequest.topBedrockIntersected),
  nationalInterest: parseNullableBooleanFilter(filterRequest.nationalInterest),
  hasGeometry: parseBooleanFilter(filterRequest.hasGeometry),
  hasLogs: parseBooleanFilter(filterRequest.hasLogs),
  hasProfiles: parseBooleanFilter(filterRequest.hasProfiles),
  hasPhotos: parseBooleanFilter(filterRequest.hasPhotos),
  hasDocuments: parseBooleanFilter(filterRequest.hasDocuments),
});

/**
 * Reads filter values from sessionStorage and constructs a FilterRequest object.
 */
export function getDefaultFilterRequestFromSession(): FilterRequest {
  const get = (key: string) => sessionStorage.getItem(key);

  const getInt = (key: string) => {
    const n = Number.parseInt(get(key) ?? "");
    return Number.isNaN(n) ? undefined : n;
  };

  const getFloat = (key: string) => {
    const n = Number.parseFloat(get(key) ?? "");
    return Number.isNaN(n) ? undefined : n;
  };

  const toArray = (n: number | undefined) => (n === undefined ? undefined : [n]);

  const allFilterParams = {
    pageNumber: (getInt(SessionKeys.page) ?? 0) + 1,
    pageSize: getInt(SessionKeys.pageSize) ?? 100,
    orderBy: get(SessionKeys.orderBy) ?? "name",
    direction: get(SessionKeys.direction) ?? "ASC",
    originalName: get(SessionKeys.originalName) ?? undefined,
    projectName: get(SessionKeys.projectName) ?? undefined,
    name: get(SessionKeys.name) ?? undefined,
    statusId: toArray(getInt(SessionKeys.statusId)),
    typeId: toArray(getInt(SessionKeys.typeId)),
    purposeId: toArray(getInt(SessionKeys.purposeId)),
    workgroupId: toArray(getInt(SessionKeys.workgroupId)),
    restrictionId: toArray(getInt(SessionKeys.restrictionId)),
    restrictionUntilFrom: get(SessionKeys.restrictionUntilFrom) ?? undefined,
    restrictionUntilTo: get(SessionKeys.restrictionUntilTo) ?? undefined,
    totalDepthMin: getFloat(SessionKeys.totalDepthMin),
    totalDepthMax: getFloat(SessionKeys.totalDepthMax),
    topBedrockFreshMdMin: getFloat(SessionKeys.topBedrockFreshMdMin),
    topBedrockFreshMdMax: getFloat(SessionKeys.topBedrockFreshMdMax),
    topBedrockWeatheredMdMin: getFloat(SessionKeys.topBedrockWeatheredMdMin),
    topBedrockWeatheredMdMax: getFloat(SessionKeys.topBedrockWeatheredMdMax),
    nationalInterest: get(SessionKeys.nationalInterest) as NullableBooleanFilterValue,
    topBedrockIntersected: get(SessionKeys.topBedrockIntersected) as NullableBooleanFilterValue,
    hasGroundwater: get(SessionKeys.hasGroundwater) as NullableBooleanFilterValue,
    hasGeometry: get(SessionKeys.hasGeometry) as BooleanFilterValue,
    hasLogs: get(SessionKeys.hasLogs) as BooleanFilterValue,
    hasProfiles: get(SessionKeys.hasProfiles) as BooleanFilterValue,
    hasPhotos: get(SessionKeys.hasPhotos) as BooleanFilterValue,
    hasDocuments: get(SessionKeys.hasDocuments) as BooleanFilterValue,
    workflowStatus: toArray(getInt(SessionKeys.workflowStatus)),
  };
  return Object.fromEntries(Object.entries(allFilterParams).filter(([, value]) => value != null));
}

export const useFilterBoreholes = (filterRequest: FilterRequest, enabled = true) => {
  const filterRequestSubmission = toFilterRequestSubmission(filterRequest);

  return useQuery({
    queryKey: [boreholeQueryKey, filterRequestSubmission],
    queryFn: () => filterBoreholes(filterRequestSubmission),
    enabled,
    placeholderData: keepPreviousData,
  });
};

export const useFilterStats = (filterRequest: FilterRequest) => {
  const filterRequestSubmission = toFilterRequestSubmission(filterRequest);

  return useQuery({
    queryKey: [boreholeQueryKey, "filter-stats", filterRequestSubmission],
    queryFn: () => fetchFilterStats(filterRequestSubmission),
    placeholderData: keepPreviousData,
  });
};

export const useBoreholeSuggestions = (field: BoreholeSuggestionField, query: string, fetchEnabled: boolean) => {
  return useQuery<BoreholeSuggestion[]>({
    queryKey: ["borehole-suggestions", field, query],
    queryFn: () => fetchBoreholeSuggestions(field, query, 10),
    enabled: fetchEnabled,
    staleTime: 30_000,
    placeholderData: previous => previous,
    retry: false,
  });
};
